src/main/java/com/example/book_your_seat/coupon/service/CouponCommandServiceImpl.java

```diff
@@ -0,0 +1,90 @@
+package com.example.book_your_seat.coupon.service;
+
+import com.example.book_your_seat.coupon.controller.dto.CouponCreateRequest;
+import com.example.book_your_seat.coupon.controller.dto.CouponIdResponse;
+import com.example.book_your_seat.coupon.domain.Coupon;
+import com.example.book_your_seat.coupon.domain.UserCoupon;
+import com.example.book_your_seat.coupon.controller.dto.UserCouponIdResponse;
+import com.example.book_your_seat.coupon.repository.CouponRepository;
+import com.example.book_your_seat.coupon.repository.UserCouponRepository;
+import com.example.book_your_seat.user.domain.User;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import static com.example.book_your_seat.coupon.CouponConst.COUPON_OUT_OF_STOCK;
+import static com.example.book_your_seat.coupon.CouponConst.DUPLICATE_BAD_REQUEST;
+
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class CouponCommandServiceImpl implements CouponCommandService {
+
+    private final CouponRepository couponRepository;
+    private final UserCouponRepository userCouponRepository;
+    private final CouponQueryService couponQueryService;
+
+    /*
+    쿠폰 발급 - 비관적 락
+     */
+    @Override
+    public UserCouponIdResponse issueCouponWithPessimistic(User user, Long couponId) {
+
+        Coupon coupon = couponQueryService.findByIdWithPessimistic(couponId);
+
+        //선착순 쿠폰 중복수령 방지
+        checkDuplicate(user.getId(), couponId);
+
+        //수량 체크
+        if(coupon.getAmount() <= 0) {
+            throw new IllegalArgumentException(COUPON_OUT_OF_STOCK);
+        }
```

Comment:
P3.5
```suggestion
        if(coupon.noStock()) {
            throw new IllegalArgumentException(COUPON_OUT_OF_STOCK);
        }
```
개인적으론 이렇게 coupon 객체에 재고가 없는지 물어보거나 
아니면 재고를 감소 시킬때 체크해주는 것이 더 객체지향적이라고 생각합니다! 

https://skyblue300a.tistory.com/18

Reply:
전 여태까지 개발 헛했네요....🥹 감사합니다 태호님 짱이십니다!!!!